### PR TITLE
Introduce `StatePath` verification to Transitions.

### PR DESCRIPTION
**Pull request by raychu86**
## Motivation

This PR introduces `inclusion_proof`s into `Transition`s - `executions` and `additional_fees`.

This proof is constructed by batching the state path verification for each input record in the transition; note that this proof must be included in each transaction if there is at least one input record. In the case that a `Transaction` does not have any input records, the `inclusion_proof` can be set to `None`.

### CODE REVIEW

File: circuit/program/src/request/verify.rs [hunk 1]
```diff
@@ -229,17 +110,43 @@ impl<A: Aleo> Request<A> {
...
+        let function_id = A::hash_bhp1024(
+            &[
+                network_id.to_bits_le(),
+                program_id.name().to_bits_le(),
+                program_id.network().to_bits_le(),
+                function_name.to_bits_le(),
+            ]
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>(),
```
Comment (ljedrz) [inline circuit/program/src/request/verify.rs]:
```suggestion
            &(
                network_id,
                program_id.name(),
                program_id.network(),
                function_name,
            )
            .to_bits_le(),
```

File: circuit/program/src/response/from_outputs.rs [hunk 2]
```diff
@@ -19,14 +19,29 @@ use super::*;
...
+        let function_id = A::hash_bhp1024(
+            &[
+                network_id.to_bits_le(),
+                program_id.name().to_bits_le(),
+                program_id.network().to_bits_le(),
+                function_name.to_bits_le(),
+            ]
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>(),
```
Comment (ljedrz) [inline circuit/program/src/response/from_outputs.rs]:
```suggestion
            &(
                network_id,
                program_id.name(),
                program_id.network(),
                function_name,
            )
            .to_bits_le(),
```

File: circuit/program/src/response/process_outputs_from_callback.rs [hunk 3]
```diff
@@ -19,13 +19,28 @@ use super::*;
...
+        let function_id = A::hash_bhp1024(
+            &[
+                network_id.to_bits_le(),
+                program_id.name().to_bits_le(),
+                program_id.network().to_bits_le(),
+                function_name.to_bits_le(),
+            ]
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>(),
```
Comment (ljedrz) [inline circuit/program/src/response/process_outputs_from_callback.rs]:
```suggestion
            &(
                network_id,
                program_id.name(),
                program_id.network(),
                function_name,
            )
            .to_bits_le(),
```

File: console/program/src/response/mod.rs [hunk 4]
```diff
@@ -51,13 +51,27 @@ impl<N: Network> Response<N> {
...
+        let function_id = N::hash_bhp1024(
+            &[
+                U16::<N>::new(N::ID).to_bits_le(),
+                program_id.name().to_bits_le(),
+                program_id.network().to_bits_le(),
+                function_name.to_bits_le(),
+            ]
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>(),
```
Comment (ljedrz) [inline console/program/src/response/mod.rs]:
```suggestion
            &(
                U16::<N>::new(N::ID),
                program_id.name(),
                program_id.network(),
                function_name,
            )
            .to_bits_le(),
```

File: console/program/src/state_path/verify.rs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            // Ensure the header path is valid.
+            ensure!(
+                N::verify_merkle_path_bhp(&self.header_path, &self.header_root, &self.header_leaf.to_bits_le()),
+                "'{}' (a header leaf) does not belong to '{}' (a block header)",
+                self.header_leaf,
+                self.block_hash
+            );
+            // Ensure the block hash is correct.
+            let preimage =
+                (*self.previous_block_hash).to_bits_le().into_iter().chain(self.header_root.to_bits_le().into_iter());
```
Comment (ljedrz) [inline console/program/src/state_path/verify.rs]:
```suggestion
                (*self.previous_block_hash).to_bits_le().into_iter().chain(self.header_root.to_bits_le());
```

File: synthesizer/src/block/transition/mod.rs [hunk 6]
```diff
@@ -83,31 +85,47 @@ impl<N: Network> Transition<N> {
...
+        let function_id = N::hash_bhp1024(
+            &[
+                network_id.to_bits_le(),
+                program_id.name().to_bits_le(),
+                program_id.network().to_bits_le(),
+                function_name.to_bits_le(),
+            ]
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>(),
```
Comment (ljedrz) [inline synthesizer/src/block/transition/mod.rs]:
```suggestion
            &(
                network_id,
                program_id.name(),
                program_id.network(),
                function_name,
            )
            .to_bits_le(),
```

File: synthesizer/src/process/additional_fee.rs [hunk 7]
```diff
@@ -87,8 +90,26 @@ impl<N: Network> Process<N> {
...
+        let function_id = N::hash_bhp1024(
+            &[
+                U16::<N>::new(N::ID).to_bits_le(),
+                additional_fee.program_id().name().to_bits_le(),
+                additional_fee.program_id().network().to_bits_le(),
+                additional_fee.function_name().to_bits_le(),
+            ]
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>(),
```
Comment (ljedrz) [inline synthesizer/src/process/additional_fee.rs]:
```suggestion
            &(
                U16::<N>::new(N::ID),
                additional_fee.program_id().name(),
                additional_fee.program_id().network(),
                additional_fee.function_name(),
            )
            .to_bits_le(),
```

File: synthesizer/src/process/execute.rs [hunk 8]
```diff
@@ -85,8 +85,26 @@ impl<N: Network> Process<N> {
...
+            let function_id = N::hash_bhp1024(
+                &[
+                    U16::<N>::new(N::ID).to_bits_le(),
+                    transition.program_id().name().to_bits_le(),
+                    transition.program_id().network().to_bits_le(),
+                    transition.function_name().to_bits_le(),
+                ]
+                .into_iter()
+                .flatten()
+                .collect::<Vec<_>>(),
```
Comment (ljedrz) [inline synthesizer/src/process/execute.rs]:
```suggestion
                &(
                    U16::<N>::new(N::ID),
                    transition.program_id().name(),
                    transition.program_id().network(),
                    transition.function_name(),
                )
                .to_bits_le(),
```

File: synthesizer/src/snark/verifying_key/mod.rs [hunk 9]
```diff
@@ -59,7 +59,7 @@ impl<N: Network> VerifyingKey<N> {
     }
 
     /// Returns `true` if the batch proof is valid for the given public inputs.
-    pub fn verify_batch(&self, function_name: &Identifier<N>, inputs: &[&[N::Field]], proof: &Proof<N>) -> bool {
+    pub fn verify_batch(&self, function_name: &str, inputs: &[Vec<N::Field>], proof: &Proof<N>) -> bool {
```
Comment (ljedrz) [inline synthesizer/src/snark/verifying_key/mod.rs:62]: what's the rationale behind this change (the `Vec`)? is this done in order for `inputs` to go out of scope at the callsite faster?

**Review by ljedrz — COMMENTED** [review]
Left some suggestions.

File: synthesizer/src/snark/verifying_key/mod.rs (hunk 9 above)
Comment (howardwu) [reply synthesizer/src/snark/verifying_key/mod.rs:62]: I had same remark. I want to see if we can return it to a slice.

File: circuit/program/Cargo.toml [hunk 10]
```diff
@@ -20,6 +20,10 @@ version = "0.9.0"
 path = "../collections"
 version = "0.9.0"
 
+[dependencies.snarkvm-circuit-environment]
```
Comment (howardwu) [inline circuit/program/Cargo.toml]: This isn't needed. See how `Request::verify` test retrieves `assert_scope`

File: console/program/src/response/mod.rs [hunk 11]
```diff
@@ -61,15 +61,7 @@ impl<N: Network> Response<N> {
...
-            &[
-                U16::<N>::new(N::ID).to_bits_le(),
-                program_id.name().to_bits_le(),
-                program_id.network().to_bits_le(),
-                function_name.to_bits_le(),
-            ]
-            .into_iter()
-            .flatten()
-            .collect::<Vec<_>>(),
+            &(U16::<N>::new(N::ID), program_id.name(), program_id.network(), function_name).to_bits_le(),
```
Comment (howardwu) [inline console/program/src/response/mod.rs]: TODO (howardwu): Change this to pass in a `network_id`.

File: synthesizer/src/snark/verifying_key/mod.rs (hunk 9 above)
Comment (raychu86) [reply synthesizer/src/snark/verifying_key/mod.rs:62]: I reverted the `verify_batch` back to slices. The one difference is that the `state_path_verifier_inputs` now uses the following logic:
`&state_path_verifier_inputs.iter().map(|inputs| inputs.as_slice()).collect::<Vec<_>>()`

Comment (howardwu) [reply synthesizer/src/snark/verifying_key/mod.rs:62]: I reverted that change, I don't think incurring a collect is faster.

**Review by howardwu — APPROVED** [review]
LGTM
